test/Swap.ts

```diff
@@ -0,0 +1,864 @@
+        buyAmount,
+        keeperStartWethBalance.sub(keeperEndWethBalance)
+      );
+
+      assert.bnEqual(
+        buyAmount,
+        userEndWethBalance.sub(userStartWethBalance)
+      );
+
+      assert.bnEqual(
+        sellAmount,
+        keeperEndUsdcBalance.sub(keeperStartUsdcBalance)
+      );
+
+      assert.bnEqual(
+        sellAmount,
+        userStartUsdcBalance.sub(userEndUsdcBalance)
+      );
+    });
+
+    it("gives the correct amount to the referrer", async function () {
+      const nonce = 1;
+      const sellAmount = totalSize.mul(minPrice).div(parseEther("1"));
+      const buyAmount = totalSize;
+      const referrer = feeRecipient;
+      const fee = 1000;
+
+      swap.setFee(feeRecipient, fee);
+
+      const userStartUsdcBalance = await usdcContract.balanceOf(user);
+      const userStartWethBalance = await wethContract.balanceOf(user);
+      const keeperStartUsdcBalance = await usdcContract.balanceOf(keeper);
+      const keeperStartWethBalance = await wethContract.balanceOf(keeper);
+      const referrerStartUsdcBalance = await usdcContract.balanceOf(referrer);
+
+      const order = {
+          swapId,
+          nonce,
+          signerWallet: user,
+          sellAmount,
+          buyAmount,
+          referrer
+        };
+
+      const signature = await getSignature(domain, order, userSigner);
+
+      const bids = [
+        [
+          swapId,
+          nonce,
+          user,
+          sellAmount,
+          buyAmount,
+          referrer,
+          signature.v,
+          signature.r,
+          signature.s,
+        ]
+      ];
+
+      const feeAmount = sellAmount.mul(fee).div(10000);
+
+      await expect(swap.connect(keeperSigner).settleOffer(
+        swapId, bids
+      )).to.emit(swap, "Swap")
+      .withArgs(
+        swapId,
+        nonce,
+        user,
+        sellAmount,
+        keeper,
+        buyAmount,
+        referrer,
+        feeAmount
+      );
+
+      const userEndUsdcBalance = await usdcContract.balanceOf(user);
+      const userEndWethBalance = await wethContract.balanceOf(user);
+      const keeperEndUsdcBalance = await usdcContract.balanceOf(keeper);
+      const keeperEndWethBalance = await wethContract.balanceOf(keeper);
+      const referrerEndUsdcBalance = await usdcContract.balanceOf(referrer);
+
+      assert.bnEqual(
+        buyAmount,
+        keeperStartWethBalance.sub(keeperEndWethBalance)
+      );
+
+      assert.bnEqual(
+        buyAmount,
+        userEndWethBalance.sub(userStartWethBalance)
+      );
+
+      assert.bnEqual(
+        sellAmount.sub(feeAmount),
+        keeperEndUsdcBalance.sub(keeperStartUsdcBalance)
+      );
+
+      assert.bnEqual(
+        sellAmount,
+        userStartUsdcBalance.sub(userEndUsdcBalance)
+      );
+
+      assert.bnEqual(
+        feeAmount,
+        referrerEndUsdcBalance.sub(referrerStartUsdcBalance)
+      );
+    });
+
+    it("fits gas budget [ @skip-on-coverage ]", async function () {
```

Comment:
Could we also have a gas benchmark for a single-bid swap? We want to optimize for that single-bid scenario